src/NServiceBus.Persistence.DynamoDB/Outbox/OutboxPersister.cs

```diff
@@ -100,15 +100,24 @@ public OutboxPersister(IAmazonDynamoDB dynamoDbClient, OutboxPersistenceConfigur
                 }
                 transportOperationsAttributes.Add(response.Items[i]);
             }
-        } while (transportOperationsAttributes?.Count < numberOfTransportOperations && response.LastEvaluatedKey.Count > 0);
+        } while (transportOperationsAttributes.Count < numberOfTransportOperations && response.LastEvaluatedKey.Count > 0);
 
-        return !foundOutboxMetadataEntry ?
-            null : DeserializeOutboxMessage(messageId, numberOfTransportOperations, transportOperationsAttributes, context);
+        if (!foundOutboxMetadataEntry)
+        {
+            return null;
+        }
+
+        if (transportOperationsAttributes!.Count != numberOfTransportOperations)
+        {
+            throw new PartialOutboxResultException(messageId, transportOperationsAttributes!.Count, numberOfTransportOperations);
+        }
+
+        return DeserializeOutboxMessage(messageId, numberOfTransportOperations, transportOperationsAttributes!, context);
     }
 
     OutboxMessage DeserializeOutboxMessage(string messageId,
         int numberOfTransportOperations,
-        List<Dictionary<string, AttributeValue>>? transportOperationsAttributes,
+        List<Dictionary<string, AttributeValue>> transportOperationsAttributes,
```

Comment:
thinking more about it, do agree that it would make more sense to check inside the method, given the method receives both the list and the number of transport operations, so it does make sense that the method validates the array access rather than relying on the data being passed to be correct.

Reply:
I'm good to go back to what I initially proposed :trollface: ;)